File: gnd/src/main/java/com/google/android/gnd/ui/util/BitmapUtil.java
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd.ui.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.provider.MediaStore.Images.Media;
import androidx.core.content.ContextCompat;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import dagger.hilt.android.qualifiers.ApplicationContext;
import java.io.IOException;
import javax.inject.Inject;

public class BitmapUtil {

  private final Context context;

  @Inject
  BitmapUtil(@ApplicationContext Context context) {
    this.context = context;
  }

  /** Retrieves an image for the given url as a {@link Bitmap}. */
  public Bitmap fromUri(Uri url) throws IOException {
    return Media.getBitmap(context.getContentResolver(), url);
  }

  public BitmapDescriptor bitmapDescriptorFromVector(int resId) {
    Drawable vectorDrawable = ContextCompat.getDrawable(context, resId);

    // Specify a bounding rectangle for the Drawable.
    vectorDrawable.setBounds(
        0, 0, vectorDrawable.getIntrinsicWidth(), vectorDrawable.getIntrinsicHeight());
    Bitmap bitmap =
        Bitmap.createBitmap(
            vectorDrawable.getIntrinsicWidth(),
            vectorDrawable.getIntrinsicHeight(),
            Bitmap.Config.ARGB_8888);
    Canvas canvas = new Canvas(bitmap);
    vectorDrawable.draw(canvas);
    return BitmapDescriptorFactory.fromBitmap(bitmap);
  }
}

File: app/src/main/java/org/groundplatform/android/ui/util/BitmapUtil.kt
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.groundplatform.android.ui.util

import android.content.Context
import android.graphics.Bitmap
import androidx.annotation.DrawableRes
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.toBitmap
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject

class BitmapUtil @Inject internal constructor(@ApplicationContext private val context: Context) {

  /**
   * Retrieves an image for the given vector resource as a [Bitmap].
   *
   * @param resId The resource ID of the vector drawable.
   * @return The decoded [Bitmap], or null if the resource is not found.
   * @throws IllegalArgumentException If the resource is not a vector drawable.
   */
  fun fromVector(@DrawableRes resId: Int): Bitmap =
    ContextCompat.getDrawable(context, resId)?.toBitmap()
      ?: throw IllegalArgumentException("Resource not found: $resId")
}
